src/librustc_typeck/check/wfcheck.rs

```diff
@@ -749,72 +748,172 @@ fn check_method_receiver<'fcx, 'gcx, 'tcx>(fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
-                .emit();
+/// returns true if `receiver_ty` would be considered a valid receiver type for `self_ty`. If
+/// `arbitrary_self_types` is enabled, `receiver_ty` must transitively deref to `self_ty`, possibly
+/// through a `*const/mut T` raw pointer. If the feature is not enabled, the requirements are more
+/// strict: `receiver_ty` must implement `Receiver` and directly implement `Deref<Target=self_ty>`.
+///
+/// NB: there are cases this function returns `true` but then causes an error to be raised later,
+/// particularly when `receiver_ty` derefs to a type that is the same as `self_ty` but has the
+/// wrong lifetime.
+fn receiver_is_valid<'fcx, 'tcx, 'gcx>(
+    fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
+    span: Span,
+    receiver_ty: Ty<'tcx>,
+    self_ty: Ty<'tcx>,
+    arbitrary_self_types_enabled: bool,
+) -> bool {
+    let cause = fcx.cause(span, traits::ObligationCauseCode::MethodReceiver);
+
+    let can_eq_self = |ty| fcx.infcx.can_eq(fcx.param_env, self_ty, ty).is_ok();
+
+    if can_eq_self(receiver_ty) {
+        if let Some(mut err) = fcx.demand_eqtype_with_origin(&cause, self_ty, receiver_ty) {
+            err.emit();
+        }
+        return true
+    }
+
+    let mut autoderef = fcx.autoderef(span, receiver_ty);
+
+    if arbitrary_self_types_enabled {
+        autoderef = autoderef.include_raw_pointers();
+    }
+
+    // skip the first type, we know its not equal to `self_ty`
+    autoderef.next();
+
+    let potential_self_ty = loop {
+        if let Some((potential_self_ty, _)) = autoderef.next() {
+            debug!("receiver_is_valid: potential self type `{:?}` to match `{:?}`",
+                potential_self_ty, self_ty);
+
+            if can_eq_self(potential_self_ty) {
+                break potential_self_ty
+            }
+        } else {
+            debug!("receiver_is_valid: type `{:?}` does not deref to `{:?}`",
+                receiver_ty, self_ty);
+            return false
+        }
+    };
+
+    if !arbitrary_self_types_enabled {
+        // check that receiver_ty: Receiver<Target=self_ty>
+
+        let receiver_trait_def_id = match fcx.tcx.lang_items().receiver_trait() {
+            Some(did) => did,
+            None => {
+                debug!("receiver_is_valid: missing Receiver trait");
+                return false
+            }
+        };
+
+        let receiver_trait_ref = ty::TraitRef{
+            def_id: receiver_trait_def_id,
+            substs: fcx.tcx.mk_substs_trait(receiver_ty, &[]),
+        };
+
+        let receiver_obligation = traits::Obligation::new(
+            cause.clone(),
+            fcx.param_env,
+            receiver_trait_ref.to_predicate()
+        );
+
+        if !fcx.predicate_must_hold(&receiver_obligation) {
+            debug!("receiver_is_valid: type `{:?}` does not implement `Receiver` trait",
+                receiver_ty);
+            return false
+        }
+
+        let deref_trait_def_id = match fcx.tcx.lang_items().deref_trait() {
```

Comment:
It'd be nice to have some comments here, to document what we are requiring and *why* (basically what's in your PR comment).

Reply:
This can reference the doc comment on the fn itself I guess.